### DISCUSSION TITLE
jsPsych Pavlovia plugin: Problem with experiment completion

### DISCUSSION THREAD

**Discussion by MaxMarschner** [jspsych/jsPsych#2686] [Q&A] [answered]
I am running an online experiment programmed with jsPsych and which I host on Pavlovia. Thus, I use the jsPsych Pavlovia plugin (2022.1.1) for communication with the Pavlovia server. Participants’ data is successfully uploaded to the server when they complete the experiment but I got feedback that on their side the experiment “freezes” after the last trial with the message “Please wait a moment while the data are uploaded to the [pavlovia.org](http://pavlovia.org/) server…”. This prevents participants to get to the redirection URL that would send them back to Prolifics where I recruit participants (causing messages about the problem, missing completion codes for participants, and potential timed-out status if participants do not close the last page themselves).

How am I supposed to embed the proper ending of the experiment within my code?

Currently it looks the following (previous lines omitted):

1. Finish connection with pavlovia

```js
 var pavlovia_finish = {
     type: jsPsychPavlovia,
     command: "finish",
     participantId: subject_id,
 };
 timeline.push(pavlovia_finish);
```

2. Insert waiting trial to ensure upload to the server is completed (as was suggested here: https://ccbrain.org/science/Onlinetesting/)

```js
 var wait5sec = {
     type: 'html-keyboard-response',
     stimulus: `<div class="message">
         <p>Wait while your data is being saved...</p>
         </div>`,
     choices: "NO_KEYS",
     trial_duration: 5000
     };

 timeline.push(wait5sec)
```

3. Redirect participants to prolific by clicking a link

```js
 var redirect_trial = {
      type: jsPsychHtmlKeyboardResponse,
      stimulus: `<div class="message">
          <p>Thank you very much for participating!</p>
          <br>
          <p><a href="https://app.prolific.co/submissions/complete?cc=XXXXXX">Click here to return to Prolific and complete the study</a>.</p>
          </div>`,
      choices: "NO_KEYS"
      }
 timeline.push(redirect_trial)
```
4. Run experiment

```js
 jsPsych.run(timeline);
```

So the problem seems to be that participants never make it to the waiting and redirect trial. Are all trials pushed after the pavlovia_finish trial omitted? Am I missing some statement in the pavlovia_finish trial (e.g., completedCallbackfunction)?

Any help or hints very appreciated!

Best
Max

**Comment by jodeleeuw** [comment]
Hi @MaxMarschner,

Do you know what the expected behavior of the Pavlovia plugin is? Is there documentation available?

**Reply by MaxMarschner** [reply]
Hi @jodeleeuw, 

so the only "documentation" I can find on this version of the plugin is the following: 
https://gitlab.pavlovia.org/shir/jsPsych_SimpleReactionTime/commit/5482642e23223b5b24417992a460f36f73824a09 --> lists changes with respect to older versions of the plugin
https://lib.pavlovia.org/jspsych-7-pavlovia-2022.1.1.js --> source code
https://gitlab.pavlovia.org/apitiot/jspsych7_simple_reaction_time/blob/master/index.html --> example script for the simple reaction time task

I think my problem is exactly that I do not know the expected behavior of the plugin... I read somewhere that I may have to add a completedCallback function statement as such: 

```
var pavlovia_finish = {
            type: jsPsychPavlovia,
            command: "finish",
            participantId: subject_id,
            completedCallback: function() {
                alert('data successfully submitted!');
            }
        };
```

Will report whether that fixed the issue. Still any hints are welcome!

Best,
Max

**Reply by becky-gilbert** [reply]
Tagging @tpronk in case he can help.

**Reply by MaxMarschner** [reply]
Just a small update. Adding the `completedCallback` statement does not solve the issue. Participants still get stuck with the page stating “Please wait a moment while the data are uploaded to the pavlovia.org server…”.

**Comment by MaxMarschner** [comment]
Another edit: I also tried to provide the redirection link within the `completedCallback` statement as such:

```
// finish connection with pavlovia.org
        var pavlovia_finish = {
            type: jsPsychPavlovia,
            command: "finish",
            participantId: subject_id,
            completedCallback: function() {
                window.location.replace("https://app.prolific.co/submissions/complete?cc=XXXXXX");
            }

        };
```

Also didn't solve the issue...

**Reply by apitiot** [reply]
Good morning,

It seems to me that the issue stems from my not expecting the pavlovia plugin to have to deal with redirection: I was under the impression that this is taken care of by jsPsych, independently from the plugin, further down the timeline.
That is to say, the pavlovia plugin only takes care of communication with the pavlovia server (it opens a session, uploads the data, and closes the session).
I am obviously happy to discuss the matter and modify the plugin accordingly. Just let me know.
Best wishes,

Alain

**Comment by seglima** [comment, marked as answer]
Hi @MaxMarschner 

I´m also using jsPsych 7 in combination with Pavlovia. I was afraid I was going to run into this problem as well but I have just tested it with Version 2021.12 (https://gitlab.pavlovia.org/shir/jsPsych_SimpleReactionTime/blob/master/jspsych-7-pavlovia-2021.12.js) and it worked just fine like this:

```
        var pavlovia_finish = {
            type: jsPsychPavlovia,
            command: "finish",
            completedCallback: function (){
                window.location.replace('https://www.google.com/');
            },
        }; 

```

Maybe using a different version of the plugin helps? Although I am not sure where this version falls in relation to the version you´re using. 

Best wishes,
Lisa

**Reply by MaxMarschner** [reply]
Hi @seglima 
thanks for your reply! 

Indeed, using the **2021.12 version** of the pavlovia plugin and providing the study completion link to the `completedCallback` statement as in your example solves my problem! 

The 2022.1.1 version of the plugin I was starting with seems to miss the completedCallback statement. I suggest using the older version (https://gitlab.pavlovia.org/shir/jsPsych_SimpleReactionTime/blob/master/jspsych-7-pavlovia-2021.12.js) until the newer version implements it. 

Best,
Max

**Reply by anshu-97** [reply]
Hi,

Just wanted to chime in and say that I was facing the same issue trying to redirect participants from Pavlovia back to Prolific using the 2022 plugin (participants feedbacked that their screen was frozen on the "Please wait a moment while the data are uploaded to the [pavlovia.org](http://pavlovia.org/) server or that their browsers crashed entirely). 

The 2021.12 version worked perfectly for me as well using the `completedCallback` parameter 👍🏻 

Thanks for the help!

**Reply by petyaracz** [reply]
When I use the 2021.12 version of the plugin with jspsych-7.1.2 I get the error `ReferenceError: $ is not defined`.  The 2022.1.1 version works but the above issue persists.

EDIT: 

For jspsych-7-pavlovia-2021.12.js you need jquery. Then it works and the problem no longer persists. Here's my index.html header:

```
<head>
	<title>Video demo [jsPsych]</title>
	<meta charset="UTF-8">

	<script src="lib/vendors/jspsych-7.1.2/jspsych.js"></script>
  <link rel="stylesheet" href="lib/vendors/jspsych-7.1.2/jspsych.css">

  <script src="lib/vendors/jspsych-7.1.2/plugin-video-keyboard-response.js"></script>
  <script src="lib/vendors/jspsych-7.1.2/plugin-image-keyboard-response.js"></script>
  <script src="lib/vendors/jspsych-7.1.2/plugin-html-keyboard-response.js"></script>
  <script src="lib/vendors/jspsych-7.1.2/plugin-preload.js"></script>
  <script type="text/javascript" src="lib/vendors/jquery-2.2.0.min.js"></script>
  <script type="text/javascript" src="plugins/jspsych-7-pavlovia-2021.12.js"></script>
  <script src="stimuli.js"></script>
  <style>
    .jspsych-button-response-button { height: 200px; width: 200px; }
  </style>

</head>
```
